python/dmdtools/batch.py

```diff
@@ -0,0 +1,469 @@
+""" Implements Dynamic Mode Decomposition, kernel DMD, exact DMD,
+    and related algorithms given sets of data
+"""
+
+import numpy as np
+# import scipy.spatial
+
+
+class DMD(object):
+    """ Dynamic Mode Decomposition (DMD)
+
+    Dynamically relevant dimensionality reduction using the Proper Orthogonal
+    Decomposition (POD) in conjunction with a least-squares solver.  This
+    approach extracts a set of modes with a fixed temporal behavior (i.e.,
+    exponential growth/decay).
+
+    This implementation uses the numpy implementation of the singular value
+    decomposition (SVD), and therefore requires the entire data set to fit in
+    memory.  This algorithm runs in O(NM^2) time, where N is the size of a
+    snapshot and M is the number of snapshots assuming N>M.
+
+    Due to the similarities in implementation, this code can compute the modes
+    associated with three variants of DMD that have appeared in the literature,
+    and a forth that is a logical combination of existing approaches:
+
+    1) Projected DMD (see Tu et al., 2014)
+    2) Exact DMD (see Tu et al., 2014)
+    3) Total least squares DMD (see Hemati & Rowley, 2015)
+    4) Projected total least squares DMD (not published, but a logical
+        combination of the Tu and Hemati papers)
+
+
+    Parameters
+    ----------
+    n_rank : int or None, optional
+        Number of POD modes to retain in the when performing DMD.  n_rank is
+        an upper bound on the rank of the resulting DMD matrix.
+
+        If n_rank is None (default), then all of the POD modes will
+        be retained.
+
+    exact : bool, optional
+        If false (default), compute the DMD modes using projected DMD
+        If true, compute the DMD modes using exact DMD
+
+        See Tu et al., 2014 for details.
+
+    total : bool, optional
+        If false (default), compute the ``standard'' DMD modes
+        If true, compute the total least squares DMD modes
+
+        See Hemati & Rowley, 2015 for details.
+
+    Properties
```

Comment:
'Properties' is not an allowed section heading in numpydoc.  Should probably be 'Attributes'